### PR TITLE
feat: Added support of indexing programs in Algolia

### PR DESCRIPTION
**Pull request by irfanuddinahmad**
## Description

This PR adds support of indexing content metadata of type 'program' into Algolia

## Ticket Link

[ENT-4865](https://openedx.atlassian.net/browse/ENT-4865)

### CODE REVIEW

File: docs/decisions/0002-celery-task-restructuring.rst [hunk 1]
```diff
@@ -25,12 +25,12 @@ There are three tasks included in this dance of updates:
 * ``update_catalog_metadata_task`` which updates the association of ``ContentMetadata`` records to
   ``CatalogQuery`` records, and does a partial update of the metadata records' ``json_metadata`` field.
 * ``update_full_content_metadata_task`` which does a full update of the metadata records' ``json_metadata``
-  based on a request to the discovery service's ``/api/v1/courses`` endpoint.
-* ``index_enterprise_catalog_courses_in_algolia_task`` which rebuilds the Algolia index with a series of
+  based on a request to the discovery service's ``/api/v1/courses`` and ``/api/v1/programs`` endpoints.
+* ``index_enterprise_catalog_in_algolia_task`` which rebuilds the Algolia index with a series of
```
Comment (HammadAhmadWaqas) [inline docs/decisions/0002-celery-task-restructuring.rst:29]: **[Question for clearity]** Can we make changes to an existing decision? Should not we add new details at the end of the file or a new file?

File: enterprise_catalog/apps/api/tasks.py [hunk 2]
```diff
@@ -443,13 +447,16 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
         # and enterprise_customer_uuids associated with each ContentMetadata record (either
-        # a course or a course run), storing them in a dictionary with the related course's
-        # content_key as a key for later retrieval. the course's content_key is determined by
-        # the content_key field if the metadata is a `COURSE` or by the parent_content_key
+        # a course or a course run or program), storing them in a dictionary with the related
+        # content_key as a key for later retrieval. the content_key is determined by
+        # the content_key field if the metadata is a `COURSE` or `PROGRAM` or by the parent_content_key
         # field if the metadata is a `COURSE_RUN`.
         for metadata in content_metadata:
             is_course_content_type = metadata.content_type == COURSE
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/api/tasks.py]: is_course_content_type is unused variable

File: enterprise_catalog/apps/api/tasks.py [hunk 3]
```diff
@@ -473,7 +480,7 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
         # if the number of uuids for both catalogs/customers exceeds ALGOLIA_UUID_BATCH_SIZE, then
         # create duplicate course records, batching the uuids (flattened records) to reduce
         # the payload size of the Algolia objects.
-        course_content_metadata = content_metadata.filter(content_type=COURSE)
+        course_content_metadata = content_metadata.filter(Q(content_type=COURSE) | Q(content_type=PROGRAM))
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/api/tasks.py]: should we need to update the variable name `course_content_metadata`?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 4]
```diff
@@ -10,9 +10,10 @@
 ALGOLIA_UUID_BATCH_SIZE = 100
 
 
-# keep attributes from course objects that we explicitly want in Algolia
+# keep attributes from contect objects that we explicitly want in Algolia
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: `content` ?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 5]
```diff
@@ -205,18 +207,19 @@ def configure_algolia_index(algolia_client):
     algolia_client.set_index_settings(ALGOLIA_INDEX_SETTINGS)
 
 
-def get_algolia_object_id(uuid):
+def get_algolia_object_id(uuid, content_type='course'):
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: Why is the course is default?

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 6]
```diff
@@ -506,50 +509,50 @@ def get_course_first_paid_enrollable_seat_price(course):
...
-        'card_image_url': get_course_card_image_url(searchable_course),
-        'advertised_course_run': get_advertised_course_run(searchable_course),
-        'skill_names': get_course_skill_names(searchable_course),
-        'skills': get_course_skills(searchable_course),
-        'first_enrollable_paid_seat_price': get_course_first_paid_enrollable_seat_price(searchable_course),
-        'original_image_url': get_course_original_image_url(searchable_course),
-        'marketing_url': get_course_marketing_url(searchable_course),
+    searchable_product = copy.deepcopy(product)
+    searchable_product.update({
+        'language': get_course_language(searchable_product),
```
Comment (HammadAhmadWaqas) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: how `get_course_language` will get language from the program product?
Same question for all the methods below.

File: enterprise_catalog/apps/api/tasks.py (hunk 2 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py]: removed

File: enterprise_catalog/apps/api/tasks.py (hunk 3 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py]: done

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 5 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: removed default

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 6 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py]: WIP: need to add program specific data ...

File: docs/decisions/0002-celery-task-restructuring.rst (hunk 1 above)
Comment (irfanuddinahmad) [reply docs/decisions/0002-celery-task-restructuring.rst:29]: I updated the task name and the method being used for courses is being replicated for programs.

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 7]
```diff
@@ -207,18 +210,19 @@ def configure_algolia_index(algolia_client):
     algolia_client.set_index_settings(ALGOLIA_INDEX_SETTINGS)
 
 
-def get_algolia_object_id(uuid):
+def get_algolia_object_id(uuid, content_type):
```
Comment (saleem-latif) [inline enterprise_catalog/apps/catalog/algolia_utils.py]: nit: I think `content_type` as first argument seems better.
```
def get_algolia_object_id(content_type, uuid):
```

**Review by saleem-latif — COMMENTED** [review]
Overall looks good so far, however, I have one question. The existing apps expect the index to have only course or course-run, will adding programs to the same index cause issues with them?
Would we need to update those places as well to filter based on content type to get the correct data?

File: enterprise_catalog/apps/api/tasks.py [hunk 8]
```diff
@@ -464,13 +468,15 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
-        # content_key as a key for later retrieval. the course's content_key is determined by
-        # the content_key field if the metadata is a `COURSE` or by the parent_content_key
+        # a course or a course run or program), storing them in a dictionary with the related
+        # content_key as a key for later retrieval. the content_key is determined by
+        # the content_key field if the metadata is a `COURSE` or `PROGRAM` or by the parent_content_key
         # field if the metadata is a `COURSE_RUN`.
         for metadata in content_metadata:
-            is_course_content_type = metadata.content_type == COURSE
-            course_content_key = metadata.content_key if is_course_content_type else metadata.parent_content_key
+            if metadata.content_type == COURSE or metadata.content_type == PROGRAM:
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py]:
```python
if metadata.content_type in (COURSE, PROGRAM):
```

File: enterprise_catalog/apps/api/tasks.py [hunk 9]
```diff
@@ -494,53 +500,53 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
     # extract out only the fields we care about and send to Algolia index
-    algolia_objects = create_algolia_objects_from_courses(courses, ALGOLIA_FIELDS)
+    algolia_objects = create_algolia_objects(products, ALGOLIA_FIELDS)
     algolia_client.replace_all_objects(algolia_objects)
 
 
 def _reindex_algolia(indexable_content_keys, nonindexable_content_keys):
     """
-    Indexes course metadata in the Algolia search index.
+    Indexes content metadata in the Algolia search index.
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py]: `Indexes course and program metadata in the Algolia search index`

File: enterprise_catalog/apps/catalog/algolia_utils.py [hunk 10]
```diff
@@ -41,7 +43,7 @@
 
 # default configuration for the index
 ALGOLIA_INDEX_SETTINGS = {
-    'attributeForDistinct': 'key',
+    'attributeForDistinct': 'aggregation_key',
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/catalog/algolia_utils.py:54]: how would this effect existing searching on course metadata?

File: enterprise_catalog/apps/api/tasks.py [hunk 11]
```diff
@@ -445,14 +449,14 @@ def index_content_keys_in_algolia(content_keys, algolia_client):
...
     )
-    courses = []
+    products = []
     for content_keys_batch in batch(content_keys, batch_size=TASK_BATCH_SIZE):
-        catalog_uuids_by_course_key = defaultdict(set)
-        customer_uuids_by_course_key = defaultdict(set)
-        catalog_queries_by_course_key = defaultdict(set)
+        catalog_uuids_by_key = defaultdict(set)
+        customer_uuids_by_key = defaultdict(set)
+        catalog_queries_by_key = defaultdict(set)
```
Comment (muhammad-ammar) [inline enterprise_catalog/apps/api/tasks.py:456]: why are we renaming these variables? existing names are more descriptive.

Comment (irfanuddinahmad) [reply enterprise_catalog/apps/api/tasks.py:456]: making them more generic to cater for both programs and courses

File: enterprise_catalog/apps/catalog/algolia_utils.py (hunk 10 above)
Comment (irfanuddinahmad) [reply enterprise_catalog/apps/catalog/algolia_utils.py:54]: No. With programs and courses, we need a common key to differentiate all the algolia objects

**Review by HammadAhmadWaqas — APPROVED** [review]

**Review by muhammad-ammar — APPROVED** [review]
LGTM

**Comment by irfanuddinahmad** [comment]
> Overall looks good so far, however, I have one question. The existing apps expect the index to have only course or course-run, will adding programs to the same index cause issues with them? Would we need to update those places as well to filter based on content type to get the correct data?

Done with https://github.com/edx/frontend-app-learner-portal-enterprise/pull/330

**Pull request merged**